### PR TITLE
[Runtime, Rest-server] Pass secret from frontend to runtime

### PR DESCRIPTION
**Pull request by Binyang2014**
Pass secret from frontend to runtime.
**Changes:**
- Create k8s secret obj and mount it for init container
- User command will not concatenate with `&&`, change to concatenate with `\n`. Runtime will write user command to a file.

**Tested:**
- job w/ & w/o secret
- job w/ & w/o private docker auth

### CODE REVIEW

File: src/kube-runtime/src/init [hunk 1]
```diff
@@ -89,6 +89,7 @@ PAI_INIT_DIR=${PAI_WORK_DIR}/init.d
 PAI_RUNTIME_DIR=${PAI_WORK_DIR}/runtime.d
 
 PAI_LOG_DIR=${PAI_WORK_DIR}/logs/${FC_POD_UID}
+PAI_SERCRET_DIR=${PAI_WORK_DIR}/secrets
```
Comment (abuccts) [inline src/kube-runtime/src/init]: typo

File: src/kube-runtime/src/init [hunk 2]
```diff
@@ -175,7 +176,8 @@ python ${PAI_INIT_DIR}/image_checker.py ${PAI_RUNTIME_DIR}/job_config.yaml
 
 # write user commands to user.sh
 # priority=100
-echo "${USER_CMD}" >> ${PAI_RUNTIME_DIR}/user.sh
+CHILD_PROCESS="RENDER_USER_COMMAND"
+python ${PAI_INIT_DIR}/user_command_renderer.py ${PAI_SERCRET_DIR}/secrets.yaml ${PAI_RUNTIME_DIR}/user.sh
```
Comment (abuccts) [inline src/kube-runtime/src/init]: typo

File: src/kube-runtime/src/init.d/user_command_renderer.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    converted_obj = {}
+    if isinstance(obj, list):
+        for i, value in enumerate(obj):
+            converted_obj[str(i)] = value
+    elif isinstance(obj, dict):
+        for key, value in obj.items():
+            converted_obj[key] = _convert_to_dict(value)
+    else:
+        converted_obj = obj
+    return converted_obj
```
Comment (abuccts) [inline src/kube-runtime/src/init.d/user_command_renderer.py]: what is this function for

File: src/kube-runtime/src/init.d/user_command_renderer.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    if not os.path.isfile(args.secret_file):
+        secrets = None
+    else:
+        with open(args.secret_file) as f:
+            secrets = yaml.safe_load(f.read())
+
+    user_command = os.getenv("USER_CMD")
+    rendered_user_command = _render_user_command(user_command, secrets)
+    _output_user_command(rendered_user_command, args.output_file)
+    logging.info("User command already rendered and output to %s",
```
Comment (abuccts) [inline src/kube-runtime/src/init.d/user_command_renderer.py]: outputted

File: src/kube-runtime/test/render_test_secrets.yaml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+time: 10
+---
+time:
+  - 10
+  - 20
+---
+time:
+  time1: 10
+  time2: 20
```
Comment (abuccts) [inline src/kube-runtime/test/render_test_secrets.yaml]: could you add some complex cases, like nested objects and arrays

File: src/kube-runtime/test/test_render_user_command.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+class TestUserCommandRender(unittest.TestCase):
+    def setUp(self):
+        try:
+            os.chdir(PACKAGE_DIRECTORY_COM)
+        except Exception:  #pylint: disable=broad-except
+            pass
+
+    def test_user_command_render(self):
+        user_commands = [
+            "sleep <% $secrets.time %>", "sleep <% $secrets.time[0] %>",
```
Comment (abuccts) [inline src/kube-runtime/test/test_render_user_command.py]: what about `$secrets.time.0`

File: src/kube-runtime/test/test_render_user_command.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+            user_commands[0], secrets[0])
+        self.assertEqual(res, "sleep 10")
+
+        res = user_command_renderer._render_user_command(
+            user_commands[1], secrets[1])
+        self.assertEqual(res, "sleep 10")
+
+        res = user_command_renderer._render_user_command(
+            user_commands[2], secrets[2])
+        self.assertEqual(res, "sleep 10")
```
Comment (abuccts) [inline src/kube-runtime/test/test_render_user_command.py]: could you use a for loop for this

File: src/rest-server/src/models/kubernetes.js [hunk 8]
```diff
@@ -85,10 +97,50 @@ const getPods = async (options = {}) => {
...
+    logger.warn('Failed to create secret, err is: ', err);
+    throw err;
+  }
+};
+
+const deleteSecret = async (namespace, name) => {
+  const client = getClient();
+  const url = `/api/v1/namespaces/${namespace}/secrets/${name}`;
+  await client.delete(url);
+  logger.debug(`delete secret ${name} successfully`);
```
Comment (abuccts) [inline src/rest-server/src/models/kubernetes.js]: do not need to add successful logs

File: src/rest-server/src/models/kubernetes.js [hunk 9]
```diff
@@ -85,10 +97,50 @@ const getPods = async (options = {}) => {
...
+};
+
+const patchSecret = async (namespace, name, data) => {
+  const client = getClient();
+  const url = `/api/v1/namespaces/${namespace}/secrets/${name}`;
+  await client.patch(url, data, patchOption);
+  logger.debug(`Patch secret ${name} successfully`);
+};
```
Comment (abuccts) [inline src/rest-server/src/models/kubernetes.js]: why do not catch errors in delete and patch but check errors in create

File: src/rest-server/src/models/kubernetes.js [hunk 10]
```diff
@@ -85,10 +97,50 @@ const getPods = async (options = {}) => {
...
+      type: type,
+    });
+    if (response.status !== status('Created')) {
+      logger.warn('Failed to create secret');
+      throw createError(response.status, 'UnknownError', response.data.message);
+    }
+    logger.debug(`create secret ${name} successfully`);
+  } catch (err) {
+    logger.warn('Failed to create secret, err is: ', err);
+    throw err;
```
Comment (abuccts) [inline src/rest-server/src/models/kubernetes.js]: pls check response body when axios response is not 20x, do not throw error directly here

File: src/rest-server/src/models/v2/job/k8s.js [hunk 11]
```diff
@@ -723,75 +735,84 @@ const createSecret = async (frameworkName, auths) => {
...
-    throw createError(response.status, 'UnknownError', response.data.message);
+};
+
+const deleteDockerSecret = async (frameworkName) => {
+  try {
+    await k8sModel
+      .getClient()
+      .deleteSecret('default', `${encodeName(frameworkName)}-regcred`);
+  } catch (error) {
+    logger.warn('Failed to delete protocol secret', error);
```
Comment (abuccts) [inline src/rest-server/src/models/v2/job/k8s.js]: remove protocol

File: src/kube-runtime/src/init.d/user_command_renderer.py (hunk 3 above)
Comment (Binyang2014) [reply src/kube-runtime/src/init.d/user_command_renderer.py]: According to issue: https://github.com/defunkt/pystache/issues/196
pystache not support syntax such as `array.0`, we need to convert array to dict first, then we can access array item.

File: src/rest-server/src/models/kubernetes.js (hunk 9 above)
Comment (Binyang2014) [reply src/rest-server/src/models/kubernetes.js]: Fixed by using axios interceptors

File: src/rest-server/src/models/kubernetes.js [hunk 12]
```diff
@@ -5,8 +5,25 @@ const axios = require('axios');
...
+const createError = require('@pai/utils/error');
 const logger = require('@pai/config/logger');
 
+const patchOption = {
+  headers: {
+    'Content-Type': 'application/merge-patch+json',
+  },
+};
+
+const rethrowResponseError = (error) => {
```
Comment (sunqinzheng) [inline src/rest-server/src/models/kubernetes.js]: I think the error handler could be enabled to each k8s requests (could be added to `getClient` function).

File: src/rest-server/src/models/v2/job/k8s.js [hunk 13]
```diff
@@ -868,7 +903,10 @@ const put = async (frameworkName, config, rawConfig) => {
   const auths = Object.values(config.prerequisites.dockerimage)
     .filter((dockerimage) => dockerimage.auth != null)
     .map((dockerimage) => dockerimage.auth);
-  auths.length && await createSecret(frameworkName, auths);
+  auths.length && await createDockerSecret(frameworkName, auths);
```
Comment (sunqinzheng) [inline src/rest-server/src/models/v2/job/k8s.js]: I think we need a more complete exception catch logic. We should store each k8s objects we created. And try to delete all created objects if error happened

File: src/rest-server/src/models/kubernetes.js (hunk 12 above)
Comment (Binyang2014) [reply src/rest-server/src/models/kubernetes.js]: As offline discuss, will leave GC to watch dog

File: src/rest-server/src/models/v2/job/k8s.js (hunk 13 above)
Comment (Binyang2014) [reply src/rest-server/src/models/v2/job/k8s.js]: As offline discuss, will leave GC to watch dog

File: src/rest-server/src/models/kubernetes.js (hunk 12 above)
Comment (sunqinzheng) [reply src/rest-server/src/models/kubernetes.js]: Ok, I was talking about another problem about axios.
Axios will throw errors with wrong stack trace and error message only conclude status code. And only this two field will be logged to console if the error is not handled.
I think we need to modify the error message and add some request info into the message (at least the request url). Another options is to log the failed request info into console and keep the original error object.

By the way, I think it is not a good idea to throw a express http error here.
The model file may be used by other models and utilities which may need the orignial axios error object with all information.

Comment (Binyang2014) [reply src/rest-server/src/models/kubernetes.js]: OK, I have changed the error handling logic. Modify the error message and add some request info into the message. Don't throw express http error here

**Pull request merged**